test/boost/reader_concurrency_semaphore_test.cc

```diff
@@ -1896,3 +1896,43 @@ SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_necessary_evicting) {
         semaphore.set_resources(initial_resources);
     }
 }
+
+// Check that a waiter permit which was queued due to the _ready_list not being
+// empty, will be executed right after the previous read in _ready_list is
+// executed, even if said read doesn't trigger admission checks via releasing
+// resources.
+SEASTAR_THREAD_TEST_CASE(test_reader_concurrency_semaphore_execution_stage_wakeup) {
+    const auto initial_resources = reader_concurrency_semaphore::resources{2, 4 * 1024};
+    const auto serialize_multiplier = std::numeric_limits<uint32_t>::max();
+    const auto kill_multiplier = std::numeric_limits<uint32_t>::max();
+    reader_concurrency_semaphore semaphore(initial_resources.count, initial_resources.memory, get_name(), 100,
+            utils::updateable_value<uint32_t>(serialize_multiplier), utils::updateable_value<uint32_t>(kill_multiplier));
+    auto stop_sem = deferred_stop(semaphore);
+
+    auto permit1 = semaphore.obtain_permit(nullptr, get_name(), 1024, db::no_timeout, {}).get();
+
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_admitted, 1);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_admitted_immediately, 1);
+
+    bool func_called = false;
+    auto func_fut = semaphore.with_ready_permit(permit1, [&] (reader_permit permit) {
+        func_called = true;
+        return sleep(std::chrono::milliseconds(1));
+    });
+    // permit1 should be on the ready list, not executed yet
+    BOOST_REQUIRE(!func_called);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 1);
+
+    // trying to obtain a second permit should block on the _ready_list
+    auto permit2_fut = semaphore.obtain_permit(nullptr, get_name(), 1024, db::no_timeout, {});
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_queued_because_ready_list, 1);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 2);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_admitted, 1);
+
+    // After func runs, the _ready_list becomes empty and the waiting permit should be admitted
+    func_fut.get();
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().waiters, 0);
+    BOOST_REQUIRE_EQUAL(semaphore.get_stats().reads_admitted, 2);
+
+    permit2_fut.get();
+}
```

Comment:
User code should not be calling `maybe_admit_waiters()`, in fact it is a private method.

Reply:
The only way I could reproduce the stalling is if the read doesn't release any release (doesn't call `signal()`). If it does, the stall doesn't reproduce, as `signal()` calls `maybe_admit_waiters()`.

Reply:
But how can that happen? Every read releases a resource (the permit itself).

Reply:
Do we call maybe_admit_waiters there? We should, since cached queries are eligible for eviction if a new query comes along.

Reply:
I think we do. on_permit_inactive() eventually calls maybe_admit_waiters().

Reply:
```c++
void reader_concurrency_semaphore::on_permit_destroyed(reader_permit::impl& permit) noexcept {
    permit.unlink();
    _permit_gate.leave();
    --_stats.current_permits;
    if (_blessed_permit == &permit) {
        _blessed_permit = nullptr;
        maybe_admit_waiters();
    }
}
```

shouldn't we admit even if we're not the blessed permit? Under non-oom conditions, there isn't a blessed permit, is there?

Reply:
So isn't this the bug? Shouldn't register_inactive_read say to itself, here I am idle, holding on to resources that could be used for an active read, shouldn't I check the queue?

Reply:
If a read is waiting on `_ready_list` to become empty, then it was already cleared w.r.t. resource availability. Dropping the inactive read will not help that waiter get admitted. In fact we used to unconditionally drop freshly registered inactive reads when there were waiters and I recently changed it because it wasted reads and lead to timeouts.

Reply:
> If a read is waiting on `_ready_list` to become empty, then it was already cleared w.r.t. resource availability. 

Yes. I guess that's *at most* one new read, and any number of resumed reads.

Sounds like we have a small but benign race:
 - move a read from _waiting to _ready
 - wake the loop
 - resume some reads, moving them to _ready and waking the loop
 - the loop awakes, but it would have been happy with just the resumed reads, we added extra load for no reason

> Dropping the inactive read will not help that waiter get admitted. In fact we used to unconditionally drop freshly registered inactive reads when there were waiters and I recently changed it because it wasted reads and lead to timeouts.

Yes. But going back to this patch. It looks like a random place to call maybe_admit_waiters(). The places to call maybe_admit_waiters() are the ones that change the inputs to its computation. Is this place changing the input to the computation?

Reply:
And indeed, 

```c++
    if (!_ready_list.empty()) {
        return {can_admit::no, reason::ready_list};
    }

```

Reply:
So I now agree with the patch, but I think _ready is too complicated, and wrong, because it creates a gap between the time we calculate whether we should start a read, and actually starting it. During that time things can change. I don't think it's critical (certainly not compared to not launching a read, as does this bug), but it makes understanding mode difficult.

Reply:
I'll queue it, though I don't like the unit test.